Tools/Scripts/libraries/webkitcorepy/webkitcorepy/multiprocessing_utils.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (C) 2022 Apple Inc. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+# 1.  Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+# 2.  Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+# ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import time
+import sys
+
+from webkitcorepy import NullContext
+
+if sys.version_info < (3, 0):
+    from multiprocessing import Lock, Pipe
+
+    import Queue as BaseQueue
+
+else:
+    from multiprocessing import context
+    from multiprocessing.queues import connection, _ForkingPickler
+
+    import queue as BaseQueue
+
+    Lock = context._default_context.Lock
+    Pipe = connection.Pipe
+
+
+# A combination of the multiprocessing module's SimpleQueue from Python 2.7.18 and 3.9.6, but with
+# a get() method that support the block and timeout arguments.
+class Queue(object):
+    Empty = BaseQueue.Empty
```

Comment:
It looks like this is a modified copy of cpython's SimpleQueue (which helps answer some questions I had about implementation details): https://github.com/python/cpython/blob/main/Lib/multiprocessing/queues.py#L337

Why copy SimpleQueue instead of subclassing it, or by creating a wrapper object that manages a SimpleQueue behind a lock? Also, IANAL but I'm not sure that we can establish copyright over this.

Reply:
Python 2 version is MIT licensed (and their license is very nearly word-for-word identical to ours), forgot to include the original author, making that change.

And I didn't sub-class `SimpleQueue` because it's kind of painful to subclass because `multiprocessing` exports `SimpleQueue` as a function. Also, different versions of Python have slightly different `SimpleQueue` implementations enough that it's easier to just do this ourselves than try and understand what is different about each Python version.  The `close()` function, for example, is new in Python 3.9.

Reply:
> Also, different versions of Python have slightly different `SimpleQueue` implementations enough that it's easier to just do this ourselves than try and understand what is different about each Python version. The `close()` function, for example, is new in Python 3.9.

Sure, but it means our implementation may easily break in future python versions.

> And I didn't sub-class `SimpleQueue` because it's kind of painful to subclass because `multiprocessing` exports `SimpleQueue` as a function.

Maybe there are details I'm missing, but I think we could use `multiprocessing.queues.SimpleQueue` directly when subclassing.

Reply:
I just spent about an hour working on this.

It works for Python 3, but not for Python 2 (because the `SimpleQueue` definition in Python 2 is kind of nonsense). I think we are going to have to spin this ourselves if we want something that is vaguely sane.

Reply:
Do we need Python 2 support here? Perhaps it's good enough to retain the existing code for Python 2, or just switch to 3 entirely.

Reply:
We could fix this bug exclusively for Python 3. The reason I didn't is because `webkitcorepy` definitely needs to support Python 2 in general, even if `run-webkit-tests` does not. It also doesn't do a ton for code complexity, it will only save us maybe 2 dozen lines (and move around the places we have to consider Python 2, we will need to, for example, bring back and conditionalize the `join_thread` calls in `TaskPool`